src/index.ts

```diff
@@ -1,65 +1,238 @@
-import fastify from 'fastify';
+import { Readable, Transform } from 'node:stream';
+import { IncomingMessage } from 'node:http';
+import fastify, {
+  FastifyInstance,
+  FastifyReply,
+  FastifyRequest,
+} from 'fastify';
 import proxy from '@fastify/http-proxy';
-import replyFrom from '@fastify/reply-from';
 import cors from '@fastify/cors';
+import FastifyMetrics from 'fastify-metrics';
 import { ENV, logger } from './util';
-import { setupShutdownHandler } from './shutdown';
+import { Shutdown } from './shutdown';
+import { isTxMulticastEnabled, performTxMulticast } from './tx-post-multicast';
+import { getCacheControlHeader } from './cache-control';
+import { startPromServer } from './metrics';
 
 const STACKS_NODE_RPC_PREFIX = '/v2';
 
-async function init() {
-  setupShutdownHandler();
+// Access-Control-Request-Method
+const ALLOW_REQUEST_METHODS = ['GET', 'PUT', 'POST', 'HEAD', 'OPTIONS'];
 
-  const server = fastify({ logger });
+// Access-Control-Request-Headers
+const ALLOW_REQUEST_HEADERS = [
+  // NOTE: unclear if routing layer will handle these for us, so for now update this list as needed
+  'Content-Type',
+  'Authorization',
+  'X-Api-Key',
+  'X-Hiro-Product',
+  'X-Hiro-Version',
+];
 
-  await server.register(cors);
+// Access-Control-Allow-Headers
+const ALLOW_RESPONSE_HEADERS = [
+  // Allow browser JS to read these response headers
+  'Content-Type',
+  'X-Api-Version',
+];
 
-  /*
-  server.addContentTypeParser(
-    'application/octet-stream',
-    (req, payload, done) => {
-      done(null, payload);
-    }
+async function postTxMulticast(
+  req: FastifyRequest,
+  reply: FastifyReply,
+  res: Readable
+) {
+  const contentType = req.headers['content-type'];
+  await performTxMulticast(req.body as Buffer, contentType);
+}
+
+/**
+ * Reads the upstream response body to memory and logs it.
+ * @param res Response from upstream
+ * @returns a new readable stream that can be used to send the response body downstream.
+ */
+function addResponseLogging(
+  req: FastifyRequest,
+  reply: FastifyReply,
+  res: Readable
+) {
+  return readPassthrough(res, (body) => {
+    const bodyString = body.toString('utf8');
+    logger.debug(
+      {
+        method: req.method,
+        url: req.url,
+        status: reply.statusCode,
+        response: bodyString,
+      },
+      `Response to ${req.method} ${req.url} returned ${reply.statusCode}: ${bodyString}`
```

Comment:
If the response body is already included in the attached object, maybe it's no longer required at the end of this string?

Reply:
I was trying to find a more "structured" log pattern, where if the transport output is json, then it _only_ emits the json, and if it's human/console, then it emits only the human-readable line, like rust's [slog](https://github.com/slog-rs/slog). It looks like Pino doesn't really support that by default, so I'm just outputting both formats :(

Could be implemented with a custom serializer or transport I think, maybe something worth trying later.